src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexQueryPlan.java

```diff
@@ -74,36 +71,108 @@ private StorageAttachedIndexQueryPlan(ColumnFamilyStore cfs,
+                                                 queryMetrics,
+                                                 rowFilter,
+                                                 ImmutableSet.copyOf(selectedIndexes),
+                                                 accumulator.complete());
+    }
 
-        for (RowFilter.Expression expression : rowFilter)
+    /**
+     * Collects the indexes that can be used with the specified filtering tree without doing a full index scan.
+     * </p>
+     * The selected indexes are those that can satisfy at least one of the expressions of the filter, and that
+     * aren't part of an OR operation that contains not indexed expressions, unless that OR operation is nested inside
+     * an AND operation that has at least one indexed operation.
+     * </p>
+     * For example, for {@code x AND y} we can use any index in {@code x}, {@code y}, or both.
+     * </p>
+     * For {@code x OR y} we can't use a single index on {@code x} or {@code y} because we would need to do a full index
+     * scan because of the unidexed expression. However, if both columns were indexed, we could use those two indexes.
+     * </p>
+     * For {@code (x OR y) AND z}, where {@code x} and {@code z} are indexed, we can use the index on {@code z}, even
+     * though we will ignore the index on {@code x}.
+     *
+     * @param element a row filter tree node
+     * @param allIndexes all the indexes in the index group
+     * @param selectedIndexes the set of indexes where we'll add those indexes can be used with the specified expression
+     * @return {@code true} if this has collected any indexes, {@code false} otherwise
+     */
+    private static boolean selectedIndexes(RowFilter.FilterElement element,
+                                           Set<StorageAttachedIndex> allIndexes,
+                                           Set<StorageAttachedIndex> selectedIndexes)
+    {
+        if (element.isDisjunction()) // OR, all restrictions should have an index
         {
-            // we ignore user-defined expressions here because we don't have a way to translate their #isSatifiedBy
-            // method, they will be included in the filter returned by QueryPlan#postIndexQueryFilter()
-            if (expression.isUserDefined())
-                continue;
-
-            acceptedExpressions.add(expression);
-            for (StorageAttachedIndex index : indexes)
+            Set<StorageAttachedIndex> orIndexes = new HashSet<>();
+            for (RowFilter.Expression expression : element.expressions())
+            {
+                if (!selectedIndexes(expression, allIndexes, orIndexes))
+                    return false;
+            }
+            for (RowFilter.FilterElement child : element.children())
+            {
+                if (!selectedIndexes(child, allIndexes, orIndexes))
+                    return false;
+            }
+            selectedIndexes.addAll(orIndexes);
+            return !orIndexes.isEmpty();
```

Comment:
could be replaced with `!element.expressions().isEmpty() || !element.children().isEmpty` to remove the `orIndexes` collection

Reply:
I'm not sure I follow. We need to collect the indexes in the deeper levels of the `OR` elements. If some of them aren't indexed, then none of the collected indexes for the previous elements in the `OR` will be used. 

If we walked the `OR` elements adding them directly to `selectedIndexes`, when we find one that doesn't match we might have already added some other indexes that we don't want to select anymore. That's why this uses a temporary set to decide find out the collected indexes should be selected once we know all have indexes.